### DISCUSSION TITLE
9.0.0-beta.8 released

### DISCUSSION THREAD

**Discussion by Feiyang1** [firebase/firebase-js-sdk#5225] [General]
`9.0.0-beta.8` has been released with the following changes:

- Updated the name of some types
	- `FirebaseFirestore` -> `Firestore`
	- `FirebaseDatabase` -> `Database`
	- `FirebaseMessaging` -> `Messaging`
	- `StorageService` -> `FirebaseStorage` (Can’t use `Storage` which conflicts with the global type for `localStorage`)
	- `Reference` (from `firebase/database`) -> `DatabaseReference`
	- `FirebaseStorageError` -> `StorageError`
	- `Settings` (from `firebase/firestore`) -> `FirestoreSettings`
	- `AnalyticsOptions` -> `AnalyticsSettings`
	- `FirebaseAppConfig` -> `FirebaseAppSettings`
	- `Settings` (from `firebase/remote-config`) -> `RemoteConfigSettings`
- `initializeApp()` becomes idempotent to make it work better with HMR and other use cases. It can be called more than once with the same options and will return the same instance. It will throw if called more than once with different options. No need to wrap `initializeApp()` in the if block anymore:

```
// the if condition can be safely removed now
// if(getApps().length === 0) {
    initializeApp({...})
// }
```
 - [Auth] Export error codes as an object map to make error code check easier
```
// ErrorCodes is a map of all possible Auth error codes
import { signInAnonymously, AuthErrorCodes } from 'firebase/auth';

async function signIn() {
    try {
    const credential = signInAnonymously(auth);
  } catch (e) {
    // instead of e.code === 'auth/network-request-failed'
    if (e.code === AuthErrorCodes.NetworkRequestFailed) {
      // do stuff
    }
  }
}
```
- [Database] Fixed https://github.com/firebase/firebase-js-sdk/issues/5195

**Comment by rileytomasek** [comment]
It looks like `ErrorCodes` should actually be `AuthErrorCodes`.
```
- import { signInAnonymously, ErrorCodes } from 'firebase/auth';
+ import { signInAnonymously, AuthErrorCodes } from 'firebase/auth';
```

**Reply by Feiyang1** [reply]
Thanks for spotting the error! I have updated the snippet to use `AuthErrorCodes`.

**Comment by mgarf** [comment]
@Feiyang1 for initializeApp being idempotent now, does that mean so is initializeAuth?

**Reply by mgarf** [reply]
here is how i refactored for reference:

```
import { useEffect, useState } from "react";
import { initializeApp } from "firebase/app"
import { initializeAuth } from "firebase/auth";


export const useFirebase = () => {
    const [firebaseApp, setFirebaseApp] = useState(null);
    const [auth, setAuth] = useState(null);
    
    useEffect(() => {
        const CLIENT_CONFIG = {
            apiKey,
            authDomain,
        };

        const appy = initializeApp(CLIENT_CONFIG);
        setFirebaseApp(appy);
        const authy = initializeAuth(appy);
        setAuth(authy);
    }, [])

    return { firebaseApp, auth }
};
```

**Reply by mgarf** [reply]
Just dug through the type documentation and its pretty clear that getApp seems like its idempotent but not initializeAuth which creates a new instance each time. Here is my rewrite which seems to be working

```
import { useEffect, useState } from "react";
import { initializeApp } from "firebase/app"
import { browserLocalPersistence, getAuth, setPersistence } from "firebase/auth";


export const useFirebase = () => {
    const [firebaseApp, setFirebaseApp] = useState(null);
    const [auth, setAuth] = useState(null);
    
    useEffect(() => {
        const CLIENT_CONFIG = {
            apiKey,
            authDomain,
        };

        const appy = initializeApp(CLIENT_CONFIG);
        const authy = getAuth(appy);

        (async () => {
            await setPersistence(authy, browserLocalPersistence);
            setFirebaseApp(appy);
            setAuth(authy);
        })();

    }, [])

    return { firebaseApp, auth }
};
```

**Reply by Feiyang1** [reply]
All `initializeFoo()`s including `initializeAuth()` are idempotent in `9.0.0`. Are you still using the beta release?

**Reply by mgarf** [reply]
hmm strange - no using 9.x.x. When i switched from initializeAuth to getAuth it started working. InitializeAuth seemed to kill the user session where as getAuth did not on page reload

`"firebase-admin": "^9.11.1"`

**Reply by Feiyang1** [reply]
`firebase-admin` is a different SDK, and is used on server. What is the version for `firebase`? Actually, can you open a new discussion thread, so it's easier to track?

BTW, this is how idempotent behavior is implemented for Auth. 
https://github.com/firebase/firebase-js-sdk/blob/bc3b405f51f5bc7a108f9c49d4b744a37756e91b/packages/auth/src/core/auth/initialize.ts#L56-L64

If it works correctly, nothing should be destroyed, because it performs nothing else other than returning the existing instance.
